### PR TITLE
feat(qa): add QA agent personalities and orchestration workflow

### PR DESCRIPTION
**Pull request by marvnramos**
## Overview

This PR introduces a complete AI-driven QA workflow built on top of Claude Code's agent system. Instead of writing test scripts manually, you define a test plan in markdown and let specialized agents execute it — one driving the UI through Playwright, another adversarially hammering the API, and an orchestrator coordinating everything in parallel.

The workflow follows a single entry point: `/qa-orchestrator`. It reads your `.qa/test-plan.md` and `.qa/config.yml`, selects the right agents, spawns them in parallel, collects results, triages bugs by severity, optionally spawns a bug-fixer for HIGH+ issues, and writes a structured report to `.qa/reports/`.

Five skills ship together:

| Skill | Role |
|---|---|
| \`qa-orchestrator\` | Main entry point — coordinates the full QA session via \`/qa-orchestrator\` |
| \`qa-happy-path\` | Drives UI flows with Playwright — navigate, fill, click, snapshot, verify |
| \`qa-chaos-monkey\` | Adversarial API tester — security, bad inputs, race conditions, edge cases |
| \`qa-bug-fixer\` | Surgical fix agent — read-first, minimal change, re-verify after fixing |
| \`qa-personality-builder\` | Guided builder for custom project-specific QA agents |

## Orchestration flow 
Here's the full orchestration flow for reference — from `/qa-orchestrator` through agent selection, parallel execution, bug triage, and report generation
```mermaid
flowchart TD
    START(["/qa-orchestrator"]):::pill
    
    START --> CONFIG["Read config files<br/>.qa/config.yml · test-plan.md · .env.qa"]:::box
    
    CONFIG --> SELECT["Select agents<br/>Playwright available? → qa-happy-path<br/>API endpoints in plan? → qa-chaos-monkey<br/>Custom personalities? → custom agents"]:::box
    
    SELECT --> CONFIRM(["Review & confirm<br/>agent selection"]):::user
    
    CONFIRM --> HP & CM & CA
    
    subgraph SPAWN ["Phase 3 — Agents run in parallel"]
        HP["qa-happy-path<br/>Navigate · fill · click · snapshot<br/>Verify network requests"]:::agent
        CM["qa-chaos-monkey<br/>Security · bad inputs · race conditions<br/>Malformed payloads · edge cases"]:::agent
        CA["Custom agent<br/>Project-specific flows"]:::agent
    end
    
    HP & CM & CA --> COLLECT["Collect results<br/>Partition → blocking bugs vs non-blocking bugs"]:::box
    
    COLLECT --> BQ{{"BLOCKER stopped<br/>a test run?"}}:::diamond
    
    BQ -- "YES — run incomplete" --> BPROMPT["BLOCKING BUG DETECTED<br/>Agent halted early · N scenarios skipped<br/>Results may be incomplete"]:::blocker
    
    BPROMPT --> BOPT(["Fix now or Abort?"]):::user
    
    BOPT -- "Fix now" --> BFX["qa-bug-fixer<br/>Fix BLOCKER only"]:::fixer
    BFX --> BRR["Re-run only<br/>failing scenarios"]:::fixer
    BRR --> BOK{{"Resolved?"}}:::diamond
    
    BOK -- "Yes" --> TRIAGE
    BOK -- "Iterate" --> BFX
    BOK -- "Give up" --> REPORT
    
    BOPT -- "Abort" --> REPORT
    
    BQ -- "NO — run complete" --> TRIAGE
    
    TRIAGE(["Non-blocking bug triage<br/>1 · Spawn bug-fixer<br/>2 · Generate report only<br/>3 · Abort"]):::user
    
    TRIAGE -- "1 · Bug-fixer" --> FX["qa-bug-fixer<br/>Fix HIGH + BLOCKER bugs"]:::fixer
    FX --> FRR["Re-run only<br/>failing scenarios"]:::fixer
    FRR --> FOK{{"Resolved?"}}:::diamond
    
    FOK -- "Yes" --> REPORT
    FOK -- "Iterate" --> FX
    FOK -- "Give up" --> REPORT
    
    TRIAGE -- "2 · Report only" --> REPORT
    TRIAGE -- "3 · Abort" --> END(["End"]):::pill
    
    REPORT["Write QA report<br/>.qa/reports/YYYY-MM-DD-HHmmss-qa-report.md"]:::reportbox
    
    REPORT --> VRD{{"Verdict"}}:::diamond
    
    VRD -- "All tests passed" --> PASS(["PASS"]):::pass
    VRD -- "Bugs remain" --> FAIL(["FAIL"]):::fail

    %% Class Definitions
    classDef pill fill:#1e1b4b,stroke:#818cf8,color:#e0e7ff,font-weight:bold,rx:99
    classDef box fill:#1e293b,stroke:#475569,color:#cbd5e1
    classDef agent fill:#0c2340,stroke:#38bdf8,color:#bae6fd,font-weight:bold
    classDef user fill:#312e81,stroke:#818cf8,color:#e0e7ff,font-weight:bold
    classDef diamond fill:#451a03,stroke:#fb923c,color:#fed7aa,font-weight:bold
    classDef blocker fill:#450a0a,stroke:#f87171,color:#fecaca,font-weight:bold
    classDef fixer fill:#431407,stroke:#fb923c,color:#ffedd5,font-weight:bold
    classDef reportbox fill:#042f2e,stroke:#2dd4bf,color:#ccfbf1
    classDef pass fill:#052e16,stroke:#4ade80,color:#bbf7d0,font-weight:bold
    classDef fail fill:#450a0a,stroke:#f87171,color:#fecaca,font-weight:bold
    
    style SPAWN fill:#0f172a,stroke:#334155,color:#94a3b8,font-style:italic
```

## Test plan

- [ ] \`ruby scripts/skills_audit.rb\` — 0 errors
- [ ] \`ruby scripts/skills_harness.rb\` — 40/40
- [ ] Install skills locally: \`cp -r skills/qa/qa-orchestrator .claude/skills/\` in a target project
- [ ] Run \`bash .claude/skills/qa-orchestrator/assets/install.sh\` — verify \`.qa/\` files generated
- [ ] Run \`/qa-orchestrator\` in Claude Code — verify mode detection and agent selection prompt
- [ ] Verify BLOCKER triage: when an agent emits \`STOPPED_EARLY: true\`, orchestrator surfaces it before non-blocking bugs

---

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by marvnramos** [comment]
# QA Report Example — 2026-04-09

**Verdict**: FAIL
**Scope**: Full test plan — Full Stack FastAPI Template
**Stack**: FastAPI + Python / PostgreSQL / React + TypeScript (Vite)
**Run date**: 2026-04-09
**API base**: http://localhost:8000/api/v1
**UI base**: http://localhost:5173 (not tested — Playwright disabled)

---

## Agent Results

| Agent | Tests | Passed | Failed | Bugs |
|---|---|---|---|---|
| qa-happy-path | skipped | — | — | — |
| qa-chaos-monkey | 91 | 86 | 5 | 3 |

> `qa-happy-path` skipped: `playwright.available: false` in `.qa/config.yml`

---

## Bug Summary

| ID | Severity | Status | Title |
|---|---|---|---|
| BUG-001 | HIGH | Fix pending deploy | 500 on `skip=-1` pagination parameter |
| BUG-002 | HIGH | Fix pending deploy | 500 on `PUT /items/{id}` with `{"title": null}` |
| BUG-003 | MEDIUM | Fix pending deploy | Invalid token returns 403 instead of 401 |

---

## Bug Details

### BUG-001 — HIGH: 500 on negative `skip` pagination parameter

- **Endpoints**: `GET /api/v1/users/`, `GET /api/v1/items/`
- **Expected**: 422 Unprocessable Entity
- **Actual**: 500 Internal Server Error
- **Root cause**: `skip` typed as `int` with no `ge=0` constraint — negative values pass validation and crash the DB layer (SQL OFFSET cannot be negative)
- **Reproduction**:
  ```bash
  curl -s "http://localhost:8000/api/v1/users/?skip=-1" \
    -H "Authorization: Bearer <admin_token>"
  # → 500 Internal Server Error
  ```
- **Fix applied**: Added `Query(default=0, ge=0)` to `skip` in:
  - `backend/app/api/routes/items.py:15`
  - `backend/app/api/routes/users.py:37`

---

### BUG-002 — HIGH: 500 on `PUT /items/{id}` with `{"title": null}`

- **Endpoint**: `PUT /api/v1/items/{id}`
- **Expected**: 422 Unprocessable Entity
- **Actual**: 500 Internal Server Error
- **Root cause**: `ItemUpdate` Pydantic schema allows `title: str | None = None` (optional update pattern), but the DB column is NOT NULL — null passes Pydantic but crashes the ORM
- **Reproduction**:
  ```bash
  curl -s -X PUT "http://localhost:8000/api/v1/items/<item_id>" \
    -H "Authorization: Bearer <token>" \
    -H "Content-Type: application/json" \
    -d '{"title": null}'
  # → 500 Internal Server Error
  ```
- **Fix applied**: Added `@field_validator("title")` to `ItemUpdate` in `backend/app/models.py:85-90` — raises `ValueError` when title is explicitly null

---

### BUG-003 — MEDIUM: Invalid token returns 403 instead of 401

- **Endpoints**: All authenticated endpoints
- **Expected**: `401 Unauthorized` + `WWW-Authenticate: Bearer` header (RFC 6750)
- **Actual**: `403 Forbidden`, no `WWW-Authenticate` header
- **Root cause**: Token validation dependency raises `HTTPException(status_code=403)` instead of `401`
- **Reproduction**:
  ```bash
  curl -s -w "\nHTTP:%{http_code}" "http://localhost:8000/api/v1/users/me" \
    -H "Authorization: Bearer garbage_token"
  # → 403 Forbidden
  ```
- **Fix applied**: Changed to `HTTP_401_UNAUTHORIZED` + `headers={"WWW-Authenticate": "Bearer"}` in `backend/app/api/deps.py:37-41`

---

## Fix Status

Fixes were implemented by `qa-bug-fixer` in an isolated worktree branch. The running application has not been restarted with the patched code — verification tests confirmed the live app still exhibits all 3 bugs.

**To apply fixes:**
1. Merge or cherry-pick the worktree branch into your working branch
2. Rebuild and restart the backend: `docker compose restart backend` (or equivalent)
3. Re-run `/qa-orchestrator` → Option C to verify

---

## Acceptance Criteria Status

| Criterion | Status |
|---|---|
| No 500 errors on any endpoint with any input | FAIL (BUG-001, BUG-002) |
| Auth enforced: 401 on missing/invalid token | FAIL (BUG-003 — returns 403) |
| 403 on unauthorized access | PASS |
| Admin-only endpoints reject regular users | PASS |
| Tenant isolation: users cannot access others' items | PASS |
| Input validation: 422 on bad data | PASS (except for null title and negative skip) |
| Password reset: no user enumeration | PASS |
| Superuser cannot delete themselves | PASS |
| Email uniqueness enforced | PASS |

---

## Recommended Next Steps

1. Apply the worktree fixes and restart the backend
2. Re-run verification: `/qa-orchestrator` → C
3. Enable Playwright (`playwright.available: true`) and re-run for UI flow coverage
---

**Comment by marvnramos** [comment]
Here's the full orchestration flow for reference — from `/qa-orchestrator` through agent selection, parallel execution, bug triage, and report generation
```mermaid
flowchart TD
    START(["/qa-orchestrator"]):::pill
    
    START --> CONFIG["Read config files<br/>.qa/config.yml · test-plan.md · .env.qa"]:::box
    
    CONFIG --> SELECT["Select agents<br/>Playwright available? → qa-happy-path<br/>API endpoints in plan? → qa-chaos-monkey<br/>Custom personalities? → custom agents"]:::box
    
    SELECT --> CONFIRM(["Review & confirm<br/>agent selection"]):::user
    
    CONFIRM --> HP & CM & CA
    
    subgraph SPAWN ["Phase 3 — Agents run in parallel"]
        HP["qa-happy-path<br/>Navigate · fill · click · snapshot<br/>Verify network requests"]:::agent
        CM["qa-chaos-monkey<br/>Security · bad inputs · race conditions<br/>Malformed payloads · edge cases"]:::agent
        CA["Custom agent<br/>Project-specific flows"]:::agent
    end
    
    HP & CM & CA --> COLLECT["Collect results<br/>Partition → blocking bugs vs non-blocking bugs"]:::box
    
    COLLECT --> BQ{{"BLOCKER stopped<br/>a test run?"}}:::diamond
    
    BQ -- "YES — run incomplete" --> BPROMPT["BLOCKING BUG DETECTED<br/>Agent halted early · N scenarios skipped<br/>Results may be incomplete"]:::blocker
    
    BPROMPT --> BOPT(["Fix now or Abort?"]):::user
    
    BOPT -- "Fix now" --> BFX["qa-bug-fixer<br/>Fix BLOCKER only"]:::fixer
    BFX --> BRR["Re-run only<br/>failing scenarios"]:::fixer
    BRR --> BOK{{"Resolved?"}}:::diamond
    
    BOK -- "Yes" --> TRIAGE
    BOK -- "Iterate" --> BFX
    BOK -- "Give up" --> REPORT
    
    BOPT -- "Abort" --> REPORT
    
    BQ -- "NO — run complete" --> TRIAGE
    
    TRIAGE(["Non-blocking bug triage<br/>1 · Spawn bug-fixer<br/>2 · Generate report only<br/>3 · Abort"]):::user
    
    TRIAGE -- "1 · Bug-fixer" --> FX["qa-bug-fixer<br/>Fix HIGH + BLOCKER bugs"]:::fixer
    FX --> FRR["Re-run only<br/>failing scenarios"]:::fixer
    FRR --> FOK{{"Resolved?"}}:::diamond
    
    FOK -- "Yes" --> REPORT
    FOK -- "Iterate" --> FX
    FOK -- "Give up" --> REPORT
    
    TRIAGE -- "2 · Report only" --> REPORT
    TRIAGE -- "3 · Abort" --> END(["End"]):::pill
    
    REPORT["Write QA report<br/>.qa/reports/YYYY-MM-DD-HHmmss-qa-report.md"]:::reportbox
    
    REPORT --> VRD{{"Verdict"}}:::diamond
    
    VRD -- "All tests passed" --> PASS(["PASS"]):::pass
    VRD -- "Bugs remain" --> FAIL(["FAIL"]):::fail

    %% Class Definitions
    classDef pill fill:#1e1b4b,stroke:#818cf8,color:#e0e7ff,font-weight:bold,rx:99
    classDef box fill:#1e293b,stroke:#475569,color:#cbd5e1
    classDef agent fill:#0c2340,stroke:#38bdf8,color:#bae6fd,font-weight:bold
    classDef user fill:#312e81,stroke:#818cf8,color:#e0e7ff,font-weight:bold
    classDef diamond fill:#451a03,stroke:#fb923c,color:#fed7aa,font-weight:bold
    classDef blocker fill:#450a0a,stroke:#f87171,color:#fecaca,font-weight:bold
    classDef fixer fill:#431407,stroke:#fb923c,color:#ffedd5,font-weight:bold
    classDef reportbox fill:#042f2e,stroke:#2dd4bf,color:#ccfbf1
    classDef pass fill:#052e16,stroke:#4ade80,color:#bbf7d0,font-weight:bold
    classDef fail fill:#450a0a,stroke:#f87171,color:#fecaca,font-weight:bold
    
    style SPAWN fill:#0f172a,stroke:#334155,color:#94a3b8,font-style:italic
```

**Comment by marvnramos** [comment]
# QA Report Example — 2026-04-09

**Verdict**: FAIL
**Scope**: Full test plan — Full Stack FastAPI Template
**Stack**: FastAPI + Python / PostgreSQL / React + TypeScript (Vite)
**Run date**: 2026-04-09
**API base**: http://localhost:8000/api/v1
**UI base**: http://localhost:5173 (not tested — Playwright disabled)

---

## Agent Results

| Agent | Tests | Passed | Failed | Bugs |
|---|---|---|---|---|
| qa-happy-path | skipped | — | — | — |
| qa-chaos-monkey | 91 | 86 | 5 | 3 |

> `qa-happy-path` skipped: `playwright.available: false` in `.qa/config.yml`

---

## Bug Summary

| ID | Severity | Status | Title |
|---|---|---|---|
| BUG-001 | HIGH | Fix pending deploy | 500 on `skip=-1` pagination parameter |
| BUG-002 | HIGH | Fix pending deploy | 500 on `PUT /items/{id}` with `{"title": null}` |
| BUG-003 | MEDIUM | Fix pending deploy | Invalid token returns 403 instead of 401 |

---

## Bug Details

### BUG-001 — HIGH: 500 on negative `skip` pagination parameter

- **Endpoints**: `GET /api/v1/users/`, `GET /api/v1/items/`
- **Expected**: 422 Unprocessable Entity
- **Actual**: 500 Internal Server Error
- **Root cause**: `skip` typed as `int` with no `ge=0` constraint — negative values pass validation and crash the DB layer (SQL OFFSET cannot be negative)
- **Reproduction**:
  ```bash
  curl -s "http://localhost:8000/api/v1/users/?skip=-1" \
    -H "Authorization: Bearer <admin_token>"
  # → 500 Internal Server Error
  ```
- **Fix applied**: Added `Query(default=0, ge=0)` to `skip` in:
  - `backend/app/api/routes/items.py:15`
  - `backend/app/api/routes/users.py:37`

---

### BUG-002 — HIGH: 500 on `PUT /items/{id}` with `{"title": null}`

- **Endpoint**: `PUT /api/v1/items/{id}`
- **Expected**: 422 Unprocessable Entity
- **Actual**: 500 Internal Server Error
- **Root cause**: `ItemUpdate` Pydantic schema allows `title: str | None = None` (optional update pattern), but the DB column is NOT NULL — null passes Pydantic but crashes the ORM
- **Reproduction**:
  ```bash
  curl -s -X PUT "http://localhost:8000/api/v1/items/<item_id>" \
    -H "Authorization: Bearer <token>" \
    -H "Content-Type: application/json" \
    -d '{"title": null}'
  # → 500 Internal Server Error
  ```
- **Fix applied**: Added `@field_validator("title")` to `ItemUpdate` in `backend/app/models.py:85-90` — raises `ValueError` when title is explicitly null

---

### BUG-003 — MEDIUM: Invalid token returns 403 instead of 401

- **Endpoints**: All authenticated endpoints
- **Expected**: `401 Unauthorized` + `WWW-Authenticate: Bearer` header (RFC 6750)
- **Actual**: `403 Forbidden`, no `WWW-Authenticate` header
- **Root cause**: Token validation dependency raises `HTTPException(status_code=403)` instead of `401`
- **Reproduction**:
  ```bash
  curl -s -w "\nHTTP:%{http_code}" "http://localhost:8000/api/v1/users/me" \
    -H "Authorization: Bearer garbage_token"
  # → 403 Forbidden
  ```
- **Fix applied**: Changed to `HTTP_401_UNAUTHORIZED` + `headers={"WWW-Authenticate": "Bearer"}` in `backend/app/api/deps.py:37-41`

---

## Fix Status

Fixes were implemented by `qa-bug-fixer` in an isolated worktree branch. The running application has not been restarted with the patched code — verification tests confirmed the live app still exhibits all 3 bugs.

**To apply fixes:**
1. Merge or cherry-pick the worktree branch into your working branch
2. Rebuild and restart the backend: `docker compose restart backend` (or equivalent)
3. Re-run `/qa-orchestrator` → Option C to verify

---

## Acceptance Criteria Status

| Criterion | Status |
|---|---|
| No 500 errors on any endpoint with any input | FAIL (BUG-001, BUG-002) |
| Auth enforced: 401 on missing/invalid token | FAIL (BUG-003 — returns 403) |
| 403 on unauthorized access | PASS |
| Admin-only endpoints reject regular users | PASS |
| Tenant isolation: users cannot access others' items | PASS |
| Input validation: 422 on bad data | PASS (except for null title and negative skip) |
| Password reset: no user enumeration | PASS |
| Superuser cannot delete themselves | PASS |
| Email uniqueness enforced | PASS |

---

## Recommended Next Steps

1. Apply the worktree fixes and restart the backend
2. Re-run verification: `/qa-orchestrator` → C
3. Enable Playwright (`playwright.available: true`) and re-run for UI flow coverage
---

File: skills/qa/qa-chaos-monkey/SKILL.md [hunk 1]
```diff
@@ -0,0 +1,165 @@
+---
+name: qa-chaos-monkey
+description: |-
+  Adversarial QA tester that systematically tries to break an application's API.
+  Tests security boundaries, input validation, race conditions, deduplication,
+  and malformed requests. Reports bugs with full reproduction details.
+  Trigger on "break the API", "chaos monkey", "adversarial testing", "security
+  test the endpoints", "test edge cases", or when a test plan defines API endpoints.
+allowed-tools: WebFetch Bash mcp__plugin_playwright_playwright__browser_navigate mcp__plugin_playwright_playwright__browser_snapshot mcp__plugin_playwright_playwright__browser_network_requests Read
```
Comment (arashiyouni) [inline skills/qa/qa-chaos-monkey/SKILL.md]: The `allowed-tools` here uses `mcp__plugin_playwright_playwright__` namespace, but the standard Claude Code Playwright MCP (installed via `claude mcp add playwright -- npx @anthropic-ai/mcp-playwright`) registers tools under `mcp__playwright__`. Users following the install instructions in `docs/qa-agents.md` will get the standard namespace, so this skill effectively won't have Playwright access. Can you confirm which setup this was tested against, or update to `mcp__playwright__browser_navigate` etc.?

File: skills/qa/qa-happy-path/SKILL.md [hunk 2]
```diff
@@ -0,0 +1,141 @@
+---
+name: qa-happy-path
+description: |-
+  Drive a web application's UI through core positive flows using Playwright browser
+  automation. Verify that primary user journeys work end-to-end, take snapshots at
+  each step, and report bugs with full reproduction details.
+  Trigger on "run happy path tests", "test the UI flows", "QA the happy path",
+  "verify the user journey", or when a test plan defines UI flows.
+allowed-tools: mcp__plugin_playwright_playwright__browser_navigate mcp__plugin_playwright_playwright__browser_click mcp__plugin_playwright_playwright__browser_fill_form mcp__plugin_playwright_playwright__browser_snapshot mcp__plugin_playwright_playwright__browser_take_screenshot mcp__plugin_playwright_playwright__browser_wait_for mcp__plugin_playwright_playwright__browser_network_requests mcp__plugin_playwright_playwright__browser_evaluate mcp__plugin_playwright_playwright__browser_type mcp__plugin_playwright_playwright__browser_press_key WebFetch
```
Comment (arashiyouni) [inline skills/qa/qa-happy-path/SKILL.md]: Same namespace mismatch as `qa-chaos-monkey` — `mcp__plugin_playwright_playwright__` vs the standard `mcp__playwright__` from `@anthropic-ai/mcp-playwright`. This skill relies heavily on Playwright so the impact here is higher.

**Review by arashiyouni — APPROVED** [review]
Overall solid PR — the orchestration concept is well-structured, the mermaid diagram is genuinely helpful, and the skill files follow repo conventions cleanly. One item worth checking before merge.

File: skills/qa/qa-chaos-monkey/SKILL.md (hunk 1 above)
Comment (marvnramos) [reply skills/qa/qa-chaos-monkey/SKILL.md]: Good catch — the skills were developed against the claude.ai desktop MCP plugin (which registers as `mcp__plugin_playwright_playwright__`), but the install docs point users to `claude mcp add playwright -- npx @anthropic-ai/mcp-playwright`, which registers as `mcp__playwright__`. Updated both `qa-chaos-monkey` and `qa-happy-path` to use the standard namespace.

File: skills/qa/qa-happy-path/SKILL.md (hunk 2 above)
Comment (marvnramos) [reply skills/qa/qa-happy-path/SKILL.md]: Fixed — same root cause as the chaos-monkey thread. Both skills now use `mcp__playwright__` to match the standard `claude mcp add playwright` install.
